contracts/distribution/subdomains/ZNSSubdomainRegistrar.sol

```diff
@@ -0,0 +1,229 @@
+                parentHash,
+                domainHash,
+                msg.sender
+            );
+        }
+
+        // TODO sub: should we clear the data from all other contracts (configs, etc.) ??
+        //  can we even do this?
+    }
+
+    function hashWithParent(
+        bytes32 parentHash,
+        string calldata label
+    ) public pure override returns (bytes32) {
+        return keccak256(
+            abi.encodePacked(
+                parentHash,
+                keccak256(bytes(label))
+            )
+        );
+    }
+
+    function setDistributionConfigForDomain(
+        bytes32 domainHash,
+        DistributionConfig calldata config
+    ) public override {
+        setPricingContractForDomain(domainHash, config.pricingContract);
+        setPaymentContractForDomain(domainHash, config.paymentContract);
+        setAccessTypeForDomain(domainHash, config.accessType);
+    }
+
+    function setPricingContractForDomain(
+        bytes32 domainHash,
+        // TODO sub: is this a problem that we expect the simplest interface
+        //  but can set any of the derived ones ??
+        AZNSPricing pricingContract
+    ) public override onlyOwnerOperatorOrRegistrar(domainHash) {
+        require(
+            address(pricingContract) != address(0),
+            "ZNSSubdomainRegistrar: pricingContract can not be 0x0 address"
+        );
+
+        distrConfigs[domainHash].pricingContract = pricingContract;
+
+        emit PricingContractSet(domainHash, address(pricingContract));
+    }
+
+    function setPaymentContractForDomain(
+        bytes32 domainHash,
+        AZNSPayment paymentContract
+    ) public override onlyOwnerOperatorOrRegistrar(domainHash) {
+        require(
+            address(paymentContract) != address(0),
+            "ZNSSubdomainRegistrar: paymentContract can not be 0x0 address"
+        );
+
+        distrConfigs[domainHash].paymentContract = AZNSPayment(paymentContract);
+        emit PaymentContractSet(domainHash, address(paymentContract));
+    }
+
+    function setAccessTypeForDomain(
+        bytes32 domainHash,
+        // TODO sub: test that we can not set the value larger
+        //  than possible values for the enum
+        AccessType accessType
+    ) public override onlyOwnerOperatorOrRegistrar(domainHash) {
+        distrConfigs[domainHash].accessType = accessType;
+        emit AccessTypeSet(domainHash, accessType);
+    }
+
+    // TODO sub: iron this out !!
+    function setWhitelistForDomain(
+        bytes32 domainHash,
+        address registrant,
+        bool allowed
+        // TODO sub: should we allow Registrar role to call this ??
+        //  if so - why ?? what can we call from there ??
+        //  should we cut Registrar out of this ??
+    ) external override onlyOwnerOperatorOrRegistrar(domainHash) {
+        distributionWhitelist[domainHash][registrant] = allowed;
+
+        emit WhitelistUpdated(domainHash, registrant, allowed);
+    }
```

Comment:
We could have a similar function but that whitelists a range of token ids and a contract address, to whitelist NFT holders.

Reply:
yes! that's a great point! need to add this

Reply:
The NFT holders that own token IDs at a contract address :)